### PR TITLE
add fast image processor nougat

### PR DESCRIPTION
**Pull request by NahieliV**
# What does this PR do?

Adds fast image processor for Nougat model. 

https://github.com/huggingface/transformers/issues/36978

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@yonigozlan

### CODE REVIEW

**Review by yonigozlan — COMMENTED** [review]
Hi @NahieliV , thanks for your contribution! Noted a few things to change :)

File: src/transformers/models/nougat/image_processing_nougat_fast.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+        return image
+
+
+    def thumbnail(
+        self,
+        image: "torch.Tensor",
+        size: SizeDict,
+        interpolation: "F.InterpolationMode" = None,
+        antialias: bool = True,
```
[the remark this answers is not in the document]
Comment (NahieliV) [reply src/transformers/models/nougat/image_processing_nougat_fast.py:253]: Removed. 
I tried to keep slow processor implementation as much as possible, and there they are passed as parameters.

File: tests/models/nougat/test_image_processing_nougat.py [hunk 2]
```diff
@@ -104,10 +109,23 @@ def prepare_image_inputs(self, equal_resolution=False, numpify=False, torchify=F
...
+    Converts a NumPy image from channels-last (H, W, C) to channels-first (C, H, W)
+    if needed. Leaves PyTorch tensors and non-NumPy types unchanged.
+    """
+    if isinstance(image, np.ndarray):
+        if image.ndim == 3 and image.shape[-1] in [1, 3]:
+            return torch.tensor(np.transpose(image, (2, 0, 1)))
+        return torch.tensor(image)
+    return image
+
+
```
[the remark this answers is not in the document]
Comment (NahieliV) [reply tests/models/nougat/test_image_processing_nougat.py:123]: I removed this for the tests of the method align_long_axis, which indeed returns the channels first if the input image is channels first. 

However,  the crop_margin method of the slow processor always returns channels lasts, even if the input image is channels first.

[the remark this answers is not in the document]
Comment (NahieliV) [reply tests/models/nougat/test_image_processing_nougat.py:123]: Furthermore, there is a bug in the original implementation when the following condition is met. 
```
 max_val = data.max()
 min_val = data.min()
```
Even when we set `data_format = "channels_first"`, the image is returned chanels last. This is because the image input format is inferred at the beginning, and then when `to_pil_image` is called, it's transformed into channels last. Then, when calling `to_channel_dimension_format`, `data_format` and `input_data_format` are the same so the function just returns the image. 

I can raise an issue and fix it if you agree.

File: src/transformers/models/nougat/image_processing_nougat_fast.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    # This generated class can be used as a starting point for the fast image processor.
+    # if the image processor is only used for simple augmentations, such as resizing, center cropping, rescaling, or normalizing,
+    # only the default values should be set in the class.
+    # If the image processor requires more complex augmentations, methods from BaseImageProcessorFast can be overridden.
+    # In most cases, only the `_preprocess` method should be overridden.
+
+    # For an example of a fast image processor requiring more complex augmentations, see `LlavaNextImageProcessorFast`.
+
+    # Default values should be checked against the slow image processor
+    # None values left after checking can be removed
```
Comment (yonigozlan) [inline src/transformers/models/nougat/image_processing_nougat_fast.py:89]:
```suggestion
```

File: src/transformers/models/nougat/image_processing_nougat_fast.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+        Crops the margin of the image. Gray pixels are considered margin (i.e., pixels with a value below the
+        threshold).
+
+        Args:
+            image (`torch.Tensor`):
+                The image to be cropped.
+            gray_threshold (`int`, *optional*, defaults to `200`)
+                Value below which pixels are considered to be gray.
+        """
+        image_rgb = image[:3, :, :]
```
Comment (yonigozlan) [inline src/transformers/models/nougat/image_processing_nougat_fast.py:160]: Is this really necessary?

File: src/transformers/models/nougat/image_processing_nougat_fast.py [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        image_std: Optional[Union[float, list[float]]],
+        return_tensors: Optional[Union[str, TensorType]],
+        **kwargs,
+    ) -> BatchFeature:
+        # Group images by size for batched resizing
+        grouped_images, grouped_images_index = group_images_by_shape(images)
+        resized_images_grouped = {}
+        for shape, stacked_images in grouped_images.items():
+            if do_crop_margin:
+                stacked_images = self._crop_margin(stacked_images)
```
Comment (yonigozlan) [inline src/transformers/models/nougat/image_processing_nougat_fast.py:336]: The output shape of this is dependent on the image itself and not just the input shape, so this can cause issue if processed by batch. Let's do this outside of a "group_images_by_shape".

File: src/transformers/models/nougat/image_processing_nougat_fast.py [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+        elif input_width > input_height:
+            height = int(input_height * width / input_width)
+
+        new_size = (height, width)
+
+        return F.resize(image, new_size, interpolation=F.InterpolationMode.BICUBIC)
+
+    def pad_image(
+        self,
+        image: "torch.Tensor",
```
Comment (yonigozlan) [inline src/transformers/models/nougat/image_processing_nougat_fast.py:253]: let's name this images and make it clear that we are padding a batch of images here

**Review by yonigozlan — COMMENTED** [review]
Thanks for iterating! Still a few changes left to make, but overall almost ready to go!

File: tests/models/nougat/test_image_processing_nougat.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yonigozlan) [reply tests/models/nougat/test_image_processing_nougat.py]: I see, nice catch! If you could fix the issue in this PR that would be great. Thanks!

File: src/transformers/models/nougat/image_processing_nougat_fast.py (hunk 4 above)
Comment (NahieliV) [reply src/transformers/models/nougat/image_processing_nougat_fast.py:160]: If the image has more than 3 channels, `F.rgb_to_grayscale` will throw an error.  This should be unlikely given that  the model's purpose  is to convert scientific PDFs to Markdown.

In the original implementation this was not necessary because `np.array(image.convert("L")).astype(np.uint8)` already does this.

**Comment by NahieliV** [comment]
@yonigozlan , changes are done. There are small differences due to the difference in the implementation of  interpolation.BICUBIC  from PIL and PyTorch. For the sample image, we have used in the test `test_slow_fast_equivalence`, there are 16 pixels with a difference over 1e-1. The max pixel difference is 0.18.

**Review by yonigozlan — COMMENTED** [review]
Hi @NahieliV ! Sorry for the delay, but thanks for iterating! Looks ready to merge to me :). Very last thing to do is to add a comment explaining why we have a larger than usual difference with slow processor, and overwrite the equivalence tests with a higher threshold.

File: src/transformers/models/nougat/image_processing_nougat_fast.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+            return image
+
+        if input_height > input_width:
+            width = int(input_width * height / input_height)
+        elif input_width > input_height:
+            height = int(input_height * width / input_width)
+
+        new_size = (height, width)
+
+        return F.resize(image, new_size, interpolation=F.InterpolationMode.BICUBIC)
```
[the remark this answers is not in the document]
Comment (NahieliV) [reply src/transformers/models/nougat/image_processing_nougat_fast.py:209]: Should be good now.

**Review by yonigozlan — APPROVED** [review]
Thanks @NahieliV for iterating on this! I had to make some small changes mainly because of recent updates in Transformers, but LGTM! Waiting for the PR to be green then I'll merge
